packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,401 @@
+            .borrow()
+            .get_state_variables()[state_var_ptr.state_var_idx]
+            .get_freshness();
+
+        // If we have cycles in the graph, it's possible that this
+        // state variable was reached from another path and already freshened.
+        // Skip it in that case.
+        match current_freshness {
+            Freshness::Fresh => continue,
+            Freshness::Unresolved => {
+                panic!("Should not have an unresolved state variable here!")
+            }
+            Freshness::Stale | Freshness::Resolved => (),
+        };
+
+        let dependencies_for_state_var =
+            &dependencies[state_var_ptr.component_idx][state_var_ptr.state_var_idx];
+
+        let mut found_stale_or_resolved_dependency = false;
+
+        for deps in dependencies_for_state_var.iter() {
+            for dep in deps.iter() {
+                match &dep.source {
+                    DependencySource::StateVar {
+                        component_idx,
+                        state_var_idx,
+                    } => {
+                        let new_state_var_ptr = StateVarPointer {
+                            component_idx: *component_idx,
+                            state_var_idx: *state_var_idx,
+                        };
+
+                        let new_current_freshness = dep.value.get_freshness();
+
+                        match new_current_freshness {
+                            // No need to recurse if the state var of the dependency is already fresh
+                            // so don't do anything to just continue the inner loop
+                            Freshness::Fresh => (),
+                            Freshness::Stale | Freshness::Resolved => {
+                                if !found_stale_or_resolved_dependency {
+                                    freshen_stack.push(state_var_ptr);
+                                    found_stale_or_resolved_dependency = true;
+                                }
+
+                                // Recurse by putting two items on the stack
+                                // and then continuing at start of stack loop
+                                freshen_stack.push(new_state_var_ptr);
+                            }
+                            Freshness::Unresolved => {
+                                panic!("How did a stale state variable depend on an unresolved state variable?")
+                            }
+                        };
+                    }
+                    // nothing to add to stack if have an essential state variable
+                    // as they cannot be stale
+                    DependencySource::Essential { .. } => (),
+                }
+            }
+        }
+
+        if !found_stale_or_resolved_dependency {
+            // All the dependencies of state_var_ptr are fresh.
+            // We calculate its value if a dependency has changed,
+            // else restore its previous value.
```

Comment:
I still don't quite get this part. We might need to talk about it.

Reply:
Changed the variable name and added more comments. Hope it is clearer now.